src/uast.cc

```diff
@@ -53,49 +73,109 @@ void *NodeAt(const Nodes *nodes, int index) {
   if (index < nodes->len) {
     return nodes->results[index];
   }
-  return NULL;
+  return nullptr;
 }
 
 Uast *UastNew(NodeIface iface) {
-  memset(error_message, 0, BUF_SIZE);
-  Uast *ctx = calloc(1, sizeof(Uast));
+  Uast *ctx;
+
+  try {
+    ctx = new Uast();
+  } catch (std::bad_alloc) {
+    Error(nullptr, "Unable to get memory\n");
+    return nullptr;
+  }
+
   if (!ctx) {
-    Error(NULL, "Unable to get memory\n");
-    return NULL;
+    Error(nullptr, "Unable to get memory\n");
+    return nullptr;
   }
   xmlInitParser();
   ctx->iface = iface;
   return ctx;
 }
 
 void UastFree(Uast *ctx) {
-  free(ctx);
+  if (ctx != nullptr)
+    delete ctx;
+
   xmlCleanupParser();
 }
 
+UastIterator *UastIteratorNew(const Uast *ctx, void *node, TreeOrder order) {
+  UastIterator *iter;
+
+  try {
+    iter = new UastIterator();
+  } catch (std::bad_alloc) {
+    Error(nullptr, "Unable to get memory\n");
+    return nullptr;
+  }
+
+  if (!iter) {
+    Error(nullptr, "Unable to get memory\n");
+    return nullptr;
+  }
+
+  iter->order = order;
+
+  iter->dqueue.push_front(node);
+  iter->ctx = ctx;
+  return iter;
+}
+
+void UastIteratorFree(UastIterator *iter) {
+  if (iter != nullptr)
+    delete iter;
+}
+
+void *UastIteratorNext(UastIterator *iter) {
+
+  if (iter == nullptr || iter->dqueue.empty()) {
+    return nullptr;
+  }
+
+  void *retNode = nullptr;
+
+  switch(iter->order) {
+    case LEVELORDER:
+      return LevelOrderNext(iter);
+    case POSTORDER:
+      return PostOrderNext(iter);
+    default:
+      return PreOrderNext(iter);
+  }
+}
+
 NodeIface UastGetIface(const Uast *ctx) { return ctx->iface; }
 
 Nodes *UastFilter(const Uast *ctx, void *node, const char *query) {
-  xmlDocPtr doc = NULL;
-  xmlXPathContextPtr xpathCtx = NULL;
-  xmlXPathObjectPtr xpathObj = NULL;
+  xmlDocPtr doc = nullptr;
+  xmlXPathContextPtr xpathCtx = nullptr;
+  xmlXPathObjectPtr xpathObj = nullptr;
   bool ok = false;
-
-  Nodes *nodes = NodesNew();
-  if (!nodes) {
-    Error(NULL, "Unable to get memory\n");
-    return NULL;
+  int size = 0;
+  xmlNodeSetPtr result = nullptr;
+  xmlNodePtr *results = nullptr;
+  xmlGenericErrorFunc handler;
+
+  Nodes *nodes;
+  try {
+    nodes = new Nodes();
+  } catch (std::bad_alloc) {
+    Error(nullptr, "Unable to get memory\n");
+    return nullptr;
   }
   doc = CreateDocument(ctx, node);
   if (!doc) {
     goto error1;
   }
-  xpathCtx = xmlXPathNewContext(doc);
+  xpathCtx = static_cast<xmlXPathContextPtr>(xmlXPathNewContext(doc));
   if (!xpathCtx) {
     goto error2;
   }
 
-  xmlGenericErrorFunc handler = (xmlGenericErrorFunc)Error;
+  handler = (xmlGenericErrorFunc)Error;
```

Comment:
Is this changed for consistency? If so, I'd rather move the declarations to where initializations are done, if possible.

Reply:
Stricter C++ rules (it doesn't allow goto jumps with declarations inside). Other options: `-fpermissive` or using an exception.